src/cas/utils/validation_utils.py

```diff
@@ -72,3 +84,263 @@ def validate_labelset_markers(annotation: Dict[str, Any], marker_list: List[str]
+        cas (Dict[str, Any]): The CAS JSON object.
+        obs (pd.DataFrame): The AnnData obs DataFrame.
+
+    Returns:
+        True if all labelset members from CAS exist in obs, otherwise False.
+    """
+    labelset_members = {}
+
+    # Collect all labelset members from CAS annotations
+    for annotation in cas[ANNOTATIONS]:
+        labelset_name = annotation[LABELSET]
+        cell_label = annotation[CELL_LABEL]
+
+        if labelset_name not in labelset_members:
+            labelset_members[labelset_name] = set()
+
+        labelset_members[labelset_name].add(cell_label)
+
+    # Check if all labelset members exist in obs
+    missing_values = {}
+
+    for labelset_name, members in labelset_members.items():
+        if labelset_name in obs.columns:
+            obs_values = set(obs[labelset_name].dropna().unique())
+            missing_members = members - obs_values
+
+            if missing_members:
+                missing_values[labelset_name] = missing_members
+
+    # Log warnings for missing values
+    if missing_values:
+        for labelset, missing in missing_values.items():
+            logging.warning(f"Missing labelset members in obs['{labelset}']: {missing}")
+        return False
+    else:
+        logging.info("All labelset members exist in the corresponding obs columns.")
+        return True
+
+
+def check_parent_child_consistency(
+    cas: Dict[str, Any], obs: Union[pd.DataFrame, CapAnnDataDF]
+) -> None:
+    """
+    Checks if the inferred hierarchy from cell labels in obs matches the expected hierarchy from
+    CAS rank data.
+
+    Parameters:
+        cas (Dict[str, Any]): The CAS JSON object containing labelset ranks.
+        obs (pd.DataFrame): The AnnData obs DataFrame.
+
+    Returns:
+        True if all inferred parent-child relationships from obs match those from cas,
+              otherwise False.
+    """
+    # Extract ranks from CAS (store expected rank per cell label)
+    cas_ranks = {entry["name"]: entry["rank"] for entry in cas[LABELSETS]}
+
+    obs_inferred_hierarchy, identical_labelset_warning = infer_obs_cell_hierarchy(
+        obs, cas_ranks
+    )
+    cas_inferred_hierarchy = infer_cas_cell_hierarchy(cas)
+
+    # Compare inferred hierarchy with CAS ranks
+    mismatches = identical_labelset_warning
+
+    for child, parent in obs_inferred_hierarchy.items():
+        if child in cas_inferred_hierarchy:
+            if parent != cas_inferred_hierarchy[child]:
+                mismatches.append(
+                    f"{child}->{parent} relation from anndata does not match with {child}->"
+                    f"{cas_inferred_hierarchy[child]} relaation from cas."
+                )
+
+    # Report mismatches
+    if mismatches:
+        for issue in mismatches:
+            logging.warning(issue)
+        return False
+    else:
+        logging.info("Parent-child relationships are consistent between CAS and OBS.")
+        return True
+
+
+def infer_obs_cell_hierarchy(
```

Comment:
it is really hard to debug this logic. Is it possible to populate a small df from a csv and write a unittest for this function